### PR TITLE
ClickHouse query hints/extensions

### PR DESCRIPTION
**Pull request by igor-tkachev**
#4066

### CODE REVIEW

File: Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+				Expression.Call(
+					null,
+					MethodHelper.GetMethodInfo(TableHint, table, hint),
+					table.Expression, Expression.Constant(hint))
+			);
+
+			return new ClickHouseSpecificTable<TSource>(newTable);
+		}
+
+//		/// <summary>
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:59]: This much commented code is auspicious in it's presence. Should it be removed, uncommented, ??  At the very least, a `// TODO:` should be added to indicate why this code is present, but commented out.

File: Source/LinqToDB/DataProvider/MySql/MySqlSqlBuilder.cs
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/MySql/MySqlSqlBuilder.cs:1]: This change has nothing to do with CH. It should be introduced on a separate PR; either one related to MySql or one specific to making this change.

Comment (igor-tkachev) [reply Source/LinqToDB/DataProvider/MySql/MySqlSqlBuilder.cs:1]: Then it is better not to do anything at all because I do not have much time for that. 
Unfortunately, we only have a choice: either we make small changes (formatting, fixing naming conventions) as we go, or we accumulate all this garbage, increase technical debt  and entropy of the universe.

**Comment by MaceWindu** [comment]
there is a buch of failing tests

File: Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+		static Expression<Func<IClickHouseSpecificTable<TSource>,IClickHouseSpecificTable<TSource>>> FinalInScopeHintImpl2<TSource>()
+			where TSource : notnull
+		{
+			return table => TableHint(table, Table.Final);
+		}
+
+		/// <summary>
+		/// Adds <b>FINAL</b> modifier to FROM Clause of all the tables in the method scope.
+		/// </summary>
+		[ExpressionMethod(ProviderName.ClickHouse, nameof(FinalInScopeHintImpl))]
```
Comment (MaceWindu) [inline Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:222]: It's not clear from description and test how it should work.
ClickHouse documentation defines FINAL as modifier to FROM clause which could be table/subquery/table_function but method description implies it will be applied to tables in subquery, which doesn't sound correct

File: Source/LinqToDB/DataProvider/MySql/MySqlSqlBuilder.cs
Comment (viceroypenguin) [reply Source/LinqToDB/DataProvider/MySql/MySqlSqlBuilder.cs:1]: You could put it on PR #4062 - it would make much more sense there.

File: Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.tt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+namespace LinqToDB.DataProvider.ClickHouse
+{
+	public static partial class ClickHouseHints
+	{
+<#
+	GenerateJoinHint("Outer");
+	GenerateJoinHint("Semi");
+	GenerateJoinHint("Anti");
+	GenerateJoinHint("Any");
+	GenerateJoinHint("AsOf");
```
[the remark this answers is not in the document]
Comment (igor-tkachev) [reply Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.tt:21]: done

File: Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs (hunk 2 above)
Comment (igor-tkachev) [reply Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:234]: This specific overload will be applied only to tables in the method scope. We do not have a method for subqueries in scope.

File: Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs [hunk 4]
```diff
@@ -21,11 +21,23 @@ public static class Table
...
+			public const string GlobalSemi  = Global + " SEMI";
+			public const string GlobalAnti  = Global + " ANTI";
+			public const string GlobalAny   = Global + " ANY";
+			public const string GlobalAsOf  = Global + " ASOF";
+			public const string All         = "ALL";
+			public const string AllOuter    = All + " OUTER";
+			public const string AllSemi     = All + " SEMI";
+			public const string AllAnti     = All + " ANTI";
+			public const string AllAny      = All + " ANY";
+			public const string AllAsOf     = All + " ASOF";
```
Comment (MaceWindu) [inline Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:40]: It doesn't make sense to me.

ClickHouse documentation defines 4 independent groups of join modifiers which doesn't match current implementation

[`[GLOBAL] [ANY|ALL|ASOF] [INNER|LEFT|RIGHT|FULL|CROSS] [OUTER|SEMI|ANTI] JOIN`](https://clickhouse.com/docs/en/sql-reference/statements/select)

Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:40]: looks like CH documentation is out-of-sync. Let me check which one is correct...

Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:40]: join is more recent, but in any case I will rise documentation issue

Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:40]: filled https://github.com/ClickHouse/clickhouse-docs/issues/1059

Comment (sdanyliv) [reply Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:40]: One note can we transform these hints from string to classes? I really don't like that extensions accepts everything that is string. I would like to see code suggestions.

Comment (igor-tkachev) [reply Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:40]: > It doesn't make sense to me.

Do you have any alternatives in mind?

File: Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs (hunk 2 above)
Comment (igor-tkachev) [reply Source/LinqToDB/DataProvider/ClickHouse/ClickHouseHints.cs:234]: We do not have such extension for now. TablesInScope was added to simplify assigning extension to all the tables in scope. The SubqueriesInScope extension is kind of ambiguous as we may generate subqueries or not depending on many reasons. If you think we really need it, I can add it.

**Comment by MaceWindu** [comment]
Could we have tests against DB to see SQL in baselines and to validate it against ClickHouse?

**Review by MaceWindu — APPROVED** [review]
